ping/ping.go

```diff
@@ -1,179 +1,124 @@
-}
+const PING = server.MessageType(1)
 
 var errPingTimedOut = errors.New("Ping took too long to response")
+var errWrongID = errors.New("the peer claims to have a different identity")
+var errFailed = errors.New("ping request failed")
 
-// Ping manages ping requests and responses
-type Ping struct {
-	p2p p2p.Service
+type Pinger struct {
+	localNode node.Node
 
-	pending    map[crypto.UUID]chan *pb.Ping
-	pendMuxtex sync.RWMutex
+	pingCallbacks []func(ping *pb.Ping) error
 
-	ingressChannel chan service.DirectMessage
+	*server.MessageServer
+	logger log.Log
+	exit   chan struct{}
 }
 
-// New creates new ping instance, receives p2p as network infra
-func New(p2p p2p.Service) *Ping {
-	p := &Ping{pending: make(map[crypto.UUID]chan *pb.Ping)}
-	p.p2p = p2p
-	p.ingressChannel = p2p.RegisterDirectProtocol(protocol)
-	go p.readLoop()
-	return p
+func (p *Pinger) RegisterCallback(f func(ping *pb.Ping) error) {
+	p.pingCallbacks = append(p.pingCallbacks, f)
 }
 
-func (p *Ping) readLoop() {
-	for {
-		msg, ok := <-p.ingressChannel
-		if !ok {
-			// Channel is closed.
-			break
-		}
-
-		go func(msg service.DirectMessage) {
-			ping := &pb.Ping{}
-			err := proto.Unmarshal(msg.Bytes(), ping)
-			if err != nil {
-				log.Error("failed to read incoming ping message err:", err)
-				// TODO : handle errors in readloop
-			}
-
-			if ping.Req {
-				log.Info("Ping: Request from (%v) - DirectMessage : %v", msg.Sender(), ping.Message)
-				err := p.handleRequest(msg.Sender(), ping)
-				if err != nil {
-					log.Error("Error handling ping request", err)
-				}
-				return
-			}
+//fires a sync every sm.syncInterval or on force space from outside
+func New(local node.Node, srv server.Service, logger log.Log) *Pinger {
+	p := Pinger{
+		localNode:     local,
+		logger:        logger,
+		MessageServer: server.NewMsgServer(srv, Protocol, PingTimeout, make(chan service.DirectMessage), logger),
+		exit:          make(chan struct{}),
+	}
 
-			p.handleResponse(ping)
+	p.RegisterMsgHandler(PING, p.newPingRequestHandler())
 
-		}(msg)
-	}
+	return &p
 }
 
-// Ping sends actual pings to target
-func (p *Ping) Ping(target p2pcrypto.PublicKey, msg string) (string, error) {
-	var response string
-	reqid := crypto.NewUUID()
-	ping := &pb.Ping{
-		ReqID:   reqid[:],
-		Req:     true,
-		Message: msg,
-	}
-	pchan, err := p.sendRequest(target, reqid, ping)
-	if err != nil {
-		return response, err
-	}
-
-	timer := time.NewTimer(PingTimeout)
-	select {
-	case res := <-pchan:
-		response = res.Message
-		p.pendMuxtex.Lock()
-		delete(p.pending, reqid)
-		p.pendMuxtex.Unlock()
-	case <-timer.C:
-		return response, errPingTimedOut
-	}
+func (p *Pinger) newPingRequestHandler() func(msg []byte) []byte {
+	return func(msg []byte) []byte {
+		p.logger.Info("handle ping request")
+		req := &pb.Ping{}
+		if err := proto.Unmarshal(msg, req); err != nil {
+			return nil
+		}
 
-	return response, nil
-}
+		for _, f := range p.pingCallbacks {
+			if err := f(req); err != nil {
```

Comment:
When a callback fails, should we not process al other callbacks? also note that this is done without paralelization and a callback might block execution of other cllbacks

Reply:
its actually like this by design.
its a base for #586. we want to make some blocking checks before replying

Reply:
ill add a note